### ISSUE TITLE
[Build] ONNX Runtime DLLs are not built with spectre mitigation

### ISSUE THREAD

**Issue by tibortakacs** [microsoft/onnxruntime#22601]
### Describe the issue

During in our build compliance check, it turned out that the Windows DLLs of the 1.19.2 release (other releases might be also affected, we have not tested them) are not built with Spectre mitigation. This can a security issue. The DLLs that we use and are affected are onnxruntime.dll, onnxruntime_providers_cuda.dll, onnxruntime_providers_shared.dll, onnxruntime_providers_tensorrt.dll.

### Target platform

Windows

### Build script

n/a

### Error / output

n/a

### Visual Studio Version

n/a

### GCC / Compiler Version

n/a

**Comment by snnn** [comment]
Where did you get the DLLs?

**Comment by snnn** [comment]
I downloaded the onnxruntime-win-x64-1.19.2.zip from our release page and ran BinSkim, here is the output:

Analyzing...
C:\Users\chasun\Downloads\onnxruntime-win-x64-1.19.2\lib\onnxruntime.dll: warning BA2024: 'onnxruntime.dll' was compiled with one or more modules that do not enable code generation mitigations for speculative execution side-channel attack (Spectre) vulnerabilities. Spectre attacks can compromise hardware-based isolation, allowing non-privileged users to retrieve potentially sensitive data from the CPU cache. To resolve the issue, provide the /Qspectre switch on the compiler command-line (or /d2guardspecload in cases where your compiler supports this switch and it is not possible to update to a toolset that supports /Qspectre). This warning should be addressed for code that operates on data that crosses a trust boundary and that can affect execution, such as parsing untrusted file inputs or processing query strings of a web request.
The following modules were compiled with a toolset that supports /Qspectre but the switch was not enabled on the command-line:
msvcprt.lib,cxx,19.40.33731.1 (filesystem.obj,locale0_implib.obj,nothrow.obj,syserror_import_lib.obj,vector_algorithms.obj,xcharconv_ryu_tables.obj,xcharconv_tables_double.obj,xcharconv_tables_float.obj,xonce2.obj)
MSVCRT.lib,cxx,19.40.33731.1 (argv_mode.obj,default_local_stdio_options.obj,delete_array.obj,delete_array_size.obj,delete_scalar.obj,delete_scalar_align.obj,delete_scalar_size.obj,delete_scalar_size_align.obj,dll_dllmain.obj,ehvecctr.obj,ehvecdtr.obj,fltused.obj,gshandler.obj,gshandlereh4.obj,initializers.obj,initsect.obj,new_array.obj,new_array_nothrow.obj,new_scalar.obj,new_scalar_align.obj,new_scalar_nothrow.obj,std_type_info_static.obj,thread_safe_statics.obj,throw_bad_alloc.obj,tlsdtor.obj,tlsdyn.obj,tlssup.obj,tncleanup.obj,ucrt_stubs.obj,utility.obj,utility_desktop.obj)
MSVCRT.lib,c,19.40.33731.1 (cpu_disp.obj,dyn_tls_init.obj,gs_cookie.obj,gs_report.obj,gs_support.obj,guard_support.obj,loadcfg.obj,ucrt_detection.obj)
vcruntime.lib,cxx,19.40.33731.1 (softmemtag.obj)


These findings are not from our code.  Is there a way to solve them?

**Comment by tibortakacs** [comment]
@snnn Sorry for the late reply. It is from the release page of GitHub.

**Comment by tibortakacs** [comment]
This output very much matches what I see during our build process. I also see that build.py script has spectre enable feature which adds /QSpectre build option. So, I am wondering if it might be straightforward to resolve this.

**Comment by snnn** [comment]
I think the log means all ORT's source code were compiled with /Qspectre, but there might be some other things we need to do.

**Comment by tjcashman** [comment]
Indeed, from https://learn.microsoft.com/en-us/cpp/build/reference/qspectre?view=msvc-170#required-libraries , it looks like we would need to also use /LIBPATH to select the Spectre-mitigated versions of the C runtime, instead of the default libraries.

**Comment by snnn** [comment]
We already enabled the property. 

![Image](https://github.com/user-attachments/assets/f5574ba6-6991-45d4-8b13-9126056d2ccc)


And we installed the Spectre libs. We installed the following Visual Studio workloads

- Microsoft.VisualStudio.Component.VC.Runtimes.x86.x64.Spectre
- Microsoft.VisualStudio.Component.VC.Tools.x86.x64

So, it should work ... I don't know what was missing.

**Comment by snnn** [comment]
Here is what I see from the latest BinSkim and the latest ONNX Runtime DLL:
```
onnxruntime.dll: warning BA2024: 'onnxruntime.dll' was compiled with one or more modules that do not enable code generation mitigations for speculative execution side-channel attack (Spectre) vulnerabilities.
Spectre attacks can compromise hardware-based isolation, allowing non-privileged users to retrieve potentially sensitive data from the CPU cache.
To resolve this issue, ensure that all modules compiled into the binary are compiled with /Qspectre switch on cl.exe command-line.
You may need to install the 'C++ spectre-mitigated libs' component from the Visual Studio installer if you observe violations against C runtime libraries such as libcmt.lib, libvcruntime.lib, etc.
For VC projects use SpectreMitigation property with 'Spectre' value.
When using older MSVC versions pass /d2guardspecload in cases where your compiler supports this switch and it is not possible to update to a toolset that supports /Qspectre.
This warning should be addressed for code that operates on data that crosses a trust boundary and that can affect execution, such as parsing untrusted file inputs or processing query strings of a web request.
For mitigation to be effective, all dlls and libs that are part of an exe, should be compiled with /QSpectre, so decision should be made for the whole process.
Can cause some perf degradations, so should not be used when not needed.
```
It still complains. But it complains nothing.
